test/helpers/manifest.go

```diff
@@ -0,0 +1,273 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package helpers
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	ginkgoext "github.com/cilium/cilium/test/ginkgo-ext"
+)
+
+// NamespaceName represents a Kubernetes namespace name
+type NamespaceName string
+
+// IsRandom returns true if the namespace name has been generated with
+// GenerateNamespaceForTest
+func (n NamespaceName) IsRandom() bool { return strings.HasPrefix(string(n), "202") }
+func (n NamespaceName) String() string { return string(n) }
+
+// Manifest represents a deployment manifest that can consist of an any number
+// of Deployments, DaemonSets, Pods, etc.
+type Manifest struct {
+	// Filename is the file (not path) of the manifest. This must point to
+	// a file that contains any number of Deployments, DaemonSets, ...
+	Filename string
+
+	// NumDaemonSet is the number of DaemonSets in the manifest
+	NumDaemonSet int
+
+	// NumPods is the number of pods expected in the manifest, not counting
+	// any DaemonSets
+	NumPods int
+
+	// LabelSelector is the selector required to select *ALL* pods created
+	// from this manifest
+	LabelSelector string
+}
+
+// Deploy deploys the manifest. It will call ginkgoext.Fail() if any aspect of
+// that fails.
+func (m Manifest) Deploy(kubectl *Kubectl, namespace string) *Deployment {
+	deploy, err := m.deploy(kubectl, namespace)
+	if err != nil {
+		ginkgoext.Failf("Unable to deploy manifest %s: %s", m.Filename, err)
+	}
+
+	return deploy
+}
+
+func (m Manifest) deploy(kubectl *Kubectl, namespace string) (*Deployment, error) {
+	ginkgoext.By("Deploying %s in namespace %s", m.Filename, namespace)
+
+	numNodes := kubectl.GetNumCiliumNodes()
+	if numNodes == 0 {
```

Comment:
Why do we need this check? It makes some sense, but on the other hand I could see possiblity of deploying a manifest which doesn't need any nodes (some crds?), so it seems weird to me to check for this here.

Reply:
It's currently used to determine the number of nodes to validate how many pods are up and ready.

The alternative would be to require listing all DaemonSet names instead of just the number of DaemonSets and then check readiness of each individual DaemonSet.